introspection-engine/connectors/mongodb-introspection-connector/src/sampler/statistics.rs

```diff
@@ -485,9 +486,86 @@ fn populate_fields(
         warnings.push(crate::warnings::fields_with_empty_names(&fields_with_empty_names));
     }
 
+    filter_out_empty_types(&mut models, &mut types, warnings);
+
     (models, types)
 }
 
+/// From the resulting data model, remove all types with no fields and change
+/// the field types to Json.
+fn filter_out_empty_types(
+    models: &mut BTreeMap<String, Model>,
+    types: &mut BTreeMap<String, CompositeType>,
+    warnings: &mut Vec<Warning>,
+) {
+    let mut fields_with_an_empty_type = Vec::new();
+
+    let empty_types: HashSet<String> = types
+        .iter()
+        .filter_map(|(name, r#type)| {
+            if r#type.fields.is_empty() {
+                Some(name.clone())
+            } else {
+                None
+            }
+        })
+        .collect();
+
+    static DOCS: &str = "Nested objects had no data in the sample dataset to introspect a nested type.";
+
+    // 1. remove all types that have no fields.
+    types.retain(|_, v| !v.fields.is_empty());
+
+    // 2. change all fields in models that point to a non-existing type to Json.
+    for (model_name, model) in models.iter_mut() {
+        let mut keep = VecDeque::with_capacity(model.fields.len());
+
+        while let Some(mut field) = model.fields.pop() {
+            if let Field::ScalarField(sf) = &mut field {
+                match &sf.field_type {
+                    datamodel::FieldType::CompositeType(ct) if !types.contains_key(ct) => {
```

Comment:
I get the logic of the condition in the context of this diff but I'm not sure it will be obvious when we come back to this code — maybe we could check if the type name is in `empty_types`?

Also less in context, but I'm wondering what is the more readable/maintainable between having a function like this that mutates the datamodel after we inferrred composite types, or avoiding to generate these in the first pass (I don't know).

Reply:
(and if we do need the vecdeque, a straightforward improvement would be creating only one and reusing it for all models, just `.clear()`ing in-between)